components/object_store/src/multipart.rs

```diff
@@ -0,0 +1,258 @@
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! Implement multipart upload of [ObjectStore](upstream::ObjectStore), and most
+//! of the codes are forked from `arrow-rs`:https://github.com/apache/arrow-rs/blob/master/object_store/src/multipart.rs
+use std::{io, pin::Pin, sync::Arc, task::Poll};
+
+use async_trait::async_trait;
+use futures::{stream::FuturesUnordered, Future, StreamExt};
+use tokio::io::AsyncWrite;
+use upstream::Result;
+
+type BoxedTryFuture<T> = Pin<Box<dyn Future<Output = Result<T, io::Error>> + Send>>;
+
+/// A trait that can be implemented by cloud-based object stores
+/// and used in combination with [`CloudMultiPartUpload`] to provide
+/// multipart upload support.
+#[async_trait]
+pub trait CloudMultiPartUploadImpl: 'static {
+    /// Upload a single part
+    async fn put_multipart_part(
+        &self,
+        buf: Vec<u8>,
+        part_idx: usize,
+    ) -> Result<UploadPart, io::Error>;
+
+    /// Complete the upload with the provided parts
+    ///
+    /// `completed_parts` is in order of part number
+    async fn complete(&self, completed_parts: Vec<UploadPart>) -> Result<(), io::Error>;
+}
+
+#[derive(Debug, Clone)]
+pub struct UploadPart {
+    pub content_id: String,
+}
+
+pub struct CloudMultiPartUpload<T>
```

Comment:
Is there any test for this? I guess we can use local file system as the `CloudMultiPartUploadImpl`. If it is not convenient, let's leave a TODO here for it.

Reply:
I guess I can add tests in pr:https://github.com/CeresDB/ceresdb/pull/887, and check every part size is equal `part_size`.